flink-staging/flink-streaming/flink-streaming-core/src/test/java/org/apache/flink/streaming/api/functions/sink/SocketClientSinkTest.java

```diff
@@ -0,0 +1,301 @@
+		}
+
+		t.join();
+		if (error.get() != null) {
+			Throwable t = error.get();
+			t.printStackTrace();
+			fail("Error in spawned thread: " + t.getMessage());
+		}
+	}
+
+	@Test
+	public void testSocketSinkRetryTenTimes() throws Exception{
+		ServerSocket server = new ServerSocket(0);
+		port = server.getLocalPort();
+
+		new Thread(new Runnable() {
+
+			@Override
+			public void run() {
+				t = Thread.currentThread();
+				SerializationSchema<String, byte[]> simpleSchema = new SerializationSchema<String, byte[]>() {
+					@Override
+					public byte[] serialize(String element) {
+						return element.getBytes();
+					}
+				};
+
+				try {
+					SocketClientSink<String> simpleSink = new SocketClientSink<String>(host, port, simpleSchema, 10);
+					simpleSink.open(new Configuration());
+
+					synchronized (t){
+						//wating for server to close
+						t.wait();
+					}
+
+					//firstly send a fin to cancel the client state:FIN_WAIT_2 while the server is CLOSE_WAIT
+					simpleSink.invoke("testSocketSinkInvoke");
+
+					//socket is closed then test "retry"
+					simpleSink.invoke("testSocketSinkInvoke");
+					simpleSink.close();
+					assertEquals(10, simpleSink.retries);
+				} catch (Exception e) {
+					//This Exception is good since the socket server is never reopen.
+				}
+			}
+		}).start();
+
+		Socket sk = server.accept();
+		sk.setKeepAlive(false);
+		sk.close();
+		server.close();
+
+		synchronized (t) {
+			t.notifyAll();
+		}
+
+		t.join();
+		if (error.get() != null) {
+			Throwable t = error.get();
+			t.printStackTrace();
+			fail("Error in spawned thread: " + t.getMessage());
+		}
+	}
+
+	@Test
+	public void testSocketSinkRetryAccess() throws Exception{
+		ServerSocket server = new ServerSocket(0);
+		port = server.getLocalPort();
+
+		new Thread(new Runnable() {
+
+			Thread t1;
+			SocketClientSink<String> simpleSink;
+
+			@Override
+			public void run() {
+				t = Thread.currentThread();
+				SerializationSchema<String, byte[]> simpleSchema = new SerializationSchema<String, byte[]>() {
+					@Override
+					public byte[] serialize(String element) {
+						return element.getBytes();
+					}
+				};
+
+				try {
+					simpleSink = new SocketClientSink<String>(host, port, simpleSchema, 10);
+					simpleSink.open(new Configuration());
+
+					synchronized (t) {
+						//wating for server to close
+						t.wait();
+					}
+
+					//firstly send a fin to cancel the client state:FIN_WAIT_2 while the server is CLOSE_WAIT
+					simpleSink.invoke("testSocketSinkInvoke");
+
+					new Thread(new Runnable() {
+
+						@Override
+						public void run() {
+							t1 = Thread.currentThread();
+							try {
+								//socket is closed then test "retry"
+								simpleSink.invoke("testSocketSinkInvoke");
+								simpleSink.close();
+							} catch (Exception e) {
+								error.set(e);
+							}
+						}
+
+					}).start();
+
+					//set a new server to let the retry success.
+					while (simpleSink.retries == 0){
```

Comment:
You're accessing shared state here from multiple threads. In this case it might happen that you don't see the updates to `simpleSink.retries` until it's too late and it has already reached the maximum number of retries.